frontend/js/src/common/brainzplayer/BrainzPlayerContext.tsx

```diff
@@ -0,0 +1,313 @@
+  type?:
+    | "SET_CURRENT_LISTEN"
+    | "SET_PLAYBACK_TIMER"
+    | "TOGGLE_REPEAT_MODE"
+    | "MOVE_QUEUE_ITEM"
+    | "MOVE_AMBIENT_QUEUE_ITEM"
+    | "REMOVE_TRACK_FROM_QUEUE"
+    | "REMOVE_TRACK_FROM_AMBIENT_QUEUE"
+    | "ADD_LISTEN_TO_TOP_OF_QUEUE"
+    | "ADD_LISTEN_TO_BOTTOM_OF_QUEUE"
+    | "ADD_LISTEN_TO_BOTTOM_OF_AMBIENT_QUEUE";
+  data?: any;
+};
+
+function valueReducer(
+  state: BrainzPlayerContextT,
+  action: BrainzPlayerActionType
+): BrainzPlayerContextT {
+  if (!action.type) {
+    return { ...state, ...action };
+  }
+
+  const isCurrentlyPlaying = (element: BrainzPlayerQueueItem) => {
+    const { currentListen } = state;
+    if (isNil(currentListen)) {
+      return false;
+    }
+
+    return isEqual(element.id, currentListen.id);
+  };
+
+  switch (action.type) {
+    case "SET_CURRENT_LISTEN": {
+      if (!action.data) {
+        return { ...state, ...action };
+      }
+      const data = action.data as BrainzPlayerQueue;
+      const { type, data: _, ...restActions } = action;
+      const newQueue = [...data].map(listenOrJSPFTrackToQueueItem);
+      if (data.length !== 0) {
+        return {
+          ...state,
+          ...restActions,
+          ambientQueue: newQueue,
+        };
+      }
+      break;
+    }
+    case "SET_PLAYBACK_TIMER": {
+      const { playerPaused, progressMs, updateTime, durationMs } = state;
+      let newProgressMs: number;
+      let elapsedTimeSinceLastUpdate: number;
+      if (playerPaused) {
+        newProgressMs = progressMs || 0;
+        elapsedTimeSinceLastUpdate = 0;
+      } else {
+        elapsedTimeSinceLastUpdate = performance.now() - updateTime;
+        const position = progressMs + elapsedTimeSinceLastUpdate;
+        newProgressMs = position > durationMs ? durationMs : position;
+      }
+      return {
+        ...state,
+        progressMs: newProgressMs,
+        updateTime: performance.now(),
+        continuousPlaybackTime:
+          state.continuousPlaybackTime + elapsedTimeSinceLastUpdate,
+      };
+    }
+    case "TOGGLE_REPEAT_MODE": {
+      const { queueRepeatMode } = state;
+      const repeatMode = repeatModes.find((mode) =>
+        isEqual(mode, queueRepeatMode)
+      );
+      const currentIndex = repeatModes.indexOf(repeatMode!);
+      const nextIndex = (currentIndex + 1) % repeatModes.length;
+      return {
+        ...state,
+        queueRepeatMode: repeatModes[nextIndex],
+      };
+    }
+    case "MOVE_QUEUE_ITEM": {
+      const { queue } = state;
+      const evt = action.data as any;
+      const currentListenIndex = queue.findIndex(isCurrentlyPlaying);
+
+      const newQueue = [...queue];
+      const newIndex = evt.newIndex + currentListenIndex;
+      const oldIndex = evt.oldIndex + currentListenIndex;
+
+      const toMove = newQueue[newIndex];
+      newQueue[newIndex] = newQueue[oldIndex];
+      newQueue[oldIndex] = toMove;
+
+      return {
+        ...state,
+        queue: newQueue,
+      };
+    }
```

Comment:
Fixed.

The issue here was that whenever the track is not playing, queue.findIndex returns -1. So this was causing an error while setting up the new queue.

Reply:
I deployed a freshly build image on test, but I still see issues when I try reordering tracks in the queue.
Wityh only 2 items in the queue it seems I can't reorder them,

With three items in the queue, reordering tracks puts them in an order different from the reordering I'm trying to do.

An example: 
Order before:
![image](https://github.com/metabrainz/listenbrainz-server/assets/6179856/c1ae6c4f-1713-4b39-9e30-41779fe49aaf)

Then moved track "Stranded" one place up, trying to swap it with "Petroleum"

Order after dropping, note that "Stranded" hasn't moved but "Petroleum" has, and it moved to position 1:
![image](https://github.com/metabrainz/listenbrainz-server/assets/6179856/6ff89ae2-5ce1-45a7-aafe-9453bf1ba768)


I put some breakpoints in the code to see why that happens, and it looks like the contents of the queue and what is displayed as the queue on screen are diverging. I suppose this is because you keep past tracks to be able to go back when pressing the back button?
Not sure how to handle this but currently there is a clear index offset issue.
Note in the console the queue with 9 items, and on the page showing 4 items in queue. I will add that the first item in this shadow queue (The Creeper) is from a previous page, and I don't know why the other ones are duplicated.
![image](https://github.com/metabrainz/listenbrainz-server/assets/6179856/af670c2f-f77e-46c5-a2e4-68e99f90e02f)